### ISSUE TITLE
Could not load python library when run from compiled julia app on another machine

### ISSUE THREAD

**Issue by sairus7** [JuliaPy/PyCall.jl#981]
When I compile my app that has a PyCall dependency, it could not find python library, because it is compiled with absolute path to that library found on dev machine. So when I run my app on another machine, it can't find it:

```
fatal: error thrown and no exception handler available.
InitError(mod=:PyCall, error=ErrorException("could not load library "C:\absolute\path\from\another\machine\.julia\conda\3\python39.dll"
The specified module could not be found. "))
```

**Comment by stevengj** [comment]
Yes, compiling in the `libpython` into PyCall was something we did to improve load times (#169).  Nowadays, it might be possible to revisit this; see e.g. the [PythonCall.jl package](https://github.com/cjdoris/PythonCall.jl) for a more dynamic approach.

**Comment by sairus7** [comment]
There are some other problems compiling app with PythonCall - see here: https://github.com/cjdoris/PythonCall.jl/issues/146

@stevengj Can you please explain a little bit more, what should be fixed in PyCall to exclude fixed library paths from compiled binaries? Or any other ways to use it with compiled app?

**Comment by efJerryYang** [comment]
I also encountered this problem today, almost the exact same error message as @sairus7 . At first I used `create_sysimage` it crashed in a very similar way, so I chose `create_app` but got this error.

**Comment by stevengj** [comment]
I wonder if it's as simple as inserting the following line [here](https://github.com/JuliaPy/PyCall.jl/blob/2728e20162db3c90198e2828c40e675924c73528/deps/build.jl#L84):
```
if 
libpython, python, pyprogramname = relpath.((libpython, python, pyprogramname), @__DIR__)
```

**Comment by sairus7** [comment]
> And I wonder if this is the right absolute path for `build.jl`?

Make sure you `dev PyCall` in your project environment and make changes into `.julia/dev/PyCall/deps/build.jl`

**Comment by sairus7** [comment]
> I build my app in default julia environment.

To compile the project into an app I use project own environment, not the default one. 
This is typical script `compile_app.sh`, notice `--project=@.` argument:
```
julia -e 'using Pkg; Pkg.add("PackageCompiler")'

julia --project=@. --startup-file=no -e 'using Pkg; Pkg.instantiate()'

julia --project=@. --startup-file=no -e '
using PackageCompiler;
PackageCompiler.create_app(pwd(), "MyProjectCompiled";
    cpu_target="generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)",
    include_transitive_dependencies=false,
    filter_stdlibs=true,
    precompile_execution_file=["test/runtests.jl"])
'
```

**Comment by sairus7** [comment]
> ERROR: LoadError: MethodError: no method matching relpath(::Ptr{Nothing}, ::String)

Maybe that's because `libpython === nothing`?

**Comment by efJerryYang** [comment]
Sorry, is this solution works for you? And that may be my problem...

**Comment by sairus7** [comment]
I will check this in a couple of hours

**Comment by efJerryYang** [comment]
Great, but problem is that the path actually should not be `nothing`, and after I commented that line pkgs could be built without error (in the `Demo` environment)

**Comment by efJerryYang** [comment]
I confirmed that the error was caused by that line of code.

after commenting that line:
```
(Demo) pkg> build
    Building GR ────→ `C:\Users\JerryYang\.julia\scratchspaces\44cfe95a-1eb2-52ea-b672-e2afdf69b78f\af237c08bda486b74318c8070adb96efa6952530\build.log`
    Building Conda ─→ `C:\Users\JerryYang\.julia\scratchspaces\44cfe95a-1eb2-52ea-b672-e2afdf69b78f\6e47d11ea2776bc5627421d59cdcc1296c058071\build.log`
    Building PyCall → `C:\Users\JerryYang\.julia\scratchspaces\44cfe95a-1eb2-52ea-b672-e2afdf69b78f\1fc929f47d7c151c839c5fc1375929766fb8edcc\build.log`
Precompiling project...
  5 dependencies successfully precompiled in 43 seconds (235 already precompiled)
```
cancel commenting that line:
```
(Demo) pkg> build
    Building GR ────→ `C:\Users\JerryYang\.julia\scratchspaces\44cfe95a-1eb2-52ea-b672-e2afdf69b78f\af237c08bda486b74318c8070adb96efa6952530\build.log`
    Building Conda ─→ `C:\Users\JerryYang\.julia\scratchspaces\44cfe95a-1eb2-52ea-b672-e2afdf69b78f\6e47d11ea2776bc5627421d59cdcc1296c058071\build.log`
    Building PyCall → `C:\Users\JerryYang\.julia\scratchspaces\44cfe95a-1eb2-52ea-b672-e2afdf69b78f\1fc929f47d7c151c839c5fc1375929766fb8edcc\build.log`
ERROR: Error building `PyCall`:
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done
...
```

**Comment by efJerryYang** [comment]
I got the answer. If that line of change can work, the code there should be `libpy_name` which is a string rather than `libpython`.

```
libpython = relpath.(libpy_name, @__DIR__)
```
And I'll try compiling my app later.

**Comment by efJerryYang** [comment]
I got the correct build result after replace `libpython` with `libpy_name` like this, which is really strange to assign a pointer type with string value. 

```julia
libpython = relpath.(libpy_name, @__DIR__)
```
The value of `libpy_name` is just the fixed absolute path `C:\absolute\path\from\another\machine\.julia\conda\3\python39.dll`

**Comment by sairus7** [comment]
> ```julia
> libpython = relpath.(libpython, @__DIR__)
> ```

This fix didn't work for me - it runs through tests and compilation stages, but trying to run the app gives the same error as in the first message.

**Comment by efJerryYang** [comment]
> it runs through tests and compilation stages, but trying to run the app gives the same error as in the first message.

It actually did not run through tests, with a skipped error message like this:
```
Precompiling project...
  3 dependencies successfully precompiled in 40 seconds (236 already precompiled, 1 skipped during auto due to previous errors)
```

**Comment by efJerryYang** [comment]
I think just replace the absolute `libpy_name` with its related path representation is not enough, because it is `libpython` that caused those errors. There may need some more changes to be made in related procedure, even not just the file `build.jl`.
